File: c/sort/bubble_sort.c
/*
 * Title    : Bubble Sort
 * Problem  : Sort N integers in ascending order
 * Category : sorting
 * Language : C
 * Author   : YeongWoong Kim
 * note     : Reads N(1~100) integers, sorts them in ascending order, and prints the result.
 */

#include <stdio.h>
#define MAX_SIZE 100   // maximum number of elements allowed

// function prototypes
void bubblesort(int* array, const int size);   // sort array in ascending order
void swap(int* element1ptr, int* element2ptr);   // swap two integer values

int main() {
    int a[MAX_SIZE];   // array to store input values
    int size;   // actual number of elements
    int i;

    // read number of elements from the user
    printf("Enter the number of elements (1 ~ %d): ", MAX_SIZE);
    scanf_s("%d", &size);

    // validate size
    if (size < 1 || size > MAX_SIZE) {
        printf("Invalid size.\n");
        return 1;
    }

    // read array values from the user
    printf("Enter %d integer values:\n", size);
    for (i = 0; i < size; i++) {
        scanf_s("%d", &a[i]);
    }

    // print original array
    printf("Data items in original order\n");
    for (i = 0; i < size; i++) {
        printf("%d\t", a[i]);
        if (i != 0 && (i + 1) % 10 == 0)
            printf("\n");   // print a newline after every 10 numbers
    }

    // sort the array
    bubblesort(a, size);

    // print sorted array
    printf("\nData items in ascending order\n");
    for (i = 0; i < size; i++) {
        printf("%d\t", a[i]);
        if (i != 0 && (i + 1) % 10 == 0)
            printf("\n");
    }

    printf("\n");
    return 0;   // successful termination
}

// sort an array of integers using the bubble sort algorithm
void bubblesort(int* array, const int size) {
    int pass;
    int j;

    // outer loop controls the number of passes
    for (pass = 0; pass < size - 1; pass++) {
        // inner loop compares adjacent elements
        for (j = 0; j < size - 1; j++) {
            // swap adjacent elements if they are out of order
            if (array[j] > array[j + 1]) {
                swap(&array[j], &array[j + 1]);
            }
        }
    } 
}

// swap the values pointed to by element1ptr and element2ptr
void swap(int* element1ptr, int* element2ptr) {
    int hold = *element1ptr;
    *element1ptr = *element2ptr;
    *element2ptr = hold;
}

File: java/sort/bubble_sort.java
/*
 * Title    : Bubble Sort
 * Problem  : Sort N integers in ascending order
 * Category : sorting
 * Language : Java
 * Author   : Yeonjun Cho
 * Note     : Reads N(1~100) integers, sorts them in ascending order, and prints the result.
 */

import java.util.Scanner;

class BubbleSort {
    static final int MAX_SIZE = 100;

    // Bubble sort with reduced inner loop (size - 1 - pass)
    static void bubbleSort(int[] array, int size) {
        for (int pass = 0; pass < size - 1; pass++) {
            for (int j = 0; j < size - 1 - pass; j++) {  // <- shrink range each pass
                if (array[j] > array[j + 1]) {
                    swap(array, j, j + 1);
                }
            }
        }
    }

    static void swap(int[] array, int i, int j) {
        int hold = array[i];
        array[i] = array[j];
        array[j] = hold;
    }

    static void printArray(int[] array, int size) {
        for (int i = 0; i < size; i++) {
            System.out.print(array[i] + "\t");
            if (i != 0 && (i + 1) % 10 == 0) System.out.println();
        }
        System.out.println();
    }

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);

        System.out.print("Enter the number of elements (1 ~ " + MAX_SIZE + "): ");
        int size = sc.nextInt();

        if (size < 1 || size > MAX_SIZE) {
            System.out.println("Invalid size.");
            sc.close();
            return;
        }

        // allocate exact size to avoid wasting space
        int[] a = new int[size];

        System.out.println("Enter " + size + " integer values:");
        for (int i = 0; i < size; i++) {
            a[i] = sc.nextInt();
        }

        System.out.println("Data items in original order");
        printArray(a, size);

        bubbleSort(a, size);

        System.out.println("\nData items in ascending order");
        printArray(a, size);

        sc.close();
    }
}
